File: src/utils/mime.ts
import { ResourceType } from '../types/index.js';

// MIME type to extension mapping
const MIME_TO_EXT: Record<string, string> = {
  'text/html': '.html',
  'text/css': '.css',
  'text/javascript': '.js',
  'application/javascript': '.js',
  'application/x-javascript': '.js',
  'text/ecmascript': '.js',
  'application/ecmascript': '.js',
  'image/png': '.png',
  'image/jpeg': '.jpg',
  'image/gif': '.gif',
  'image/webp': '.webp',
  'image/avif': '.avif',
  'image/svg+xml': '.svg',
  'image/x-icon': '.ico',
  'image/vnd.microsoft.icon': '.ico',
  'image/bmp': '.bmp',
  'font/woff': '.woff',
  'font/woff2': '.woff2',
  'font/ttf': '.ttf',
  'font/otf': '.otf',
  'application/font-woff': '.woff',
  'application/font-woff2': '.woff2',
  'application/x-font-ttf': '.ttf',
  'application/x-font-otf': '.otf',
  'application/vnd.ms-fontobject': '.eot',
  'video/mp4': '.mp4',
  'video/webm': '.webm',
  'video/ogg': '.ogg',
  'audio/mpeg': '.mp3',
  'audio/wav': '.wav',
  'audio/ogg': '.ogg',
  'audio/flac': '.flac',
  'application/pdf': '.pdf',
  'application/json': '.json',
  'application/manifest+json': '.webmanifest',
  'application/xml': '.xml',
  'text/xml': '.xml',
  'text/plain': '.txt',
};

// Extension to MIME type mapping
const EXT_TO_MIME: Record<string, string> = {
  '.html': 'text/html',
  '.htm': 'text/html',
  '.xhtml': 'application/xhtml+xml',
  '.css': 'text/css',
  '.js': 'application/javascript',
  '.mjs': 'application/javascript',
  '.json': 'application/json',
  '.xml': 'application/xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
  '.avif': 'image/avif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.bmp': 'image/bmp',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.otf': 'font/otf',
  '.eot': 'application/vnd.ms-fontobject',
  '.mp4': 'video/mp4',
  '.webm': 'video/webm',
  '.ogg': 'video/ogg',
  '.mp3': 'audio/mpeg',
  '.wav': 'audio/wav',
  '.flac': 'audio/flac',
  '.pdf': 'application/pdf',
  '.txt': 'text/plain',
  '.webmanifest': 'application/manifest+json',
};

// Extension to ResourceType mapping
const EXT_TO_TYPE: Record<string, ResourceType> = {
  '.html': 'html',
  '.htm': 'html',
  '.xhtml': 'html',
  '.css': 'css',
  '.js': 'javascript',
  '.mjs': 'javascript',
  '.ts': 'javascript',
  '.jsx': 'javascript',
  '.tsx': 'javascript',
  '.png': 'image',
  '.jpg': 'image',
  '.jpeg': 'image',
  '.gif': 'image',
  '.webp': 'image',
  '.avif': 'image',
  '.svg': 'image',
  '.ico': 'image',
  '.bmp': 'image',
  '.woff': 'font',
  '.woff2': 'font',
  '.ttf': 'font',
  '.otf': 'font',
  '.eot': 'font',
  '.mp4': 'video',
  '.webm': 'video',
  '.ogg': 'video',
  '.mp3': 'audio',
  '.wav': 'audio',
  '.flac': 'audio',
  '.pdf': 'document',
  '.json': 'manifest',
  '.webmanifest': 'manifest',
};

/**
 * Get file extension from MIME type
 */
export function mimeToExtension(mimeType: string): string {
  // Remove charset and other params
  const baseMime = mimeType.split(';')[0]?.trim().toLowerCase() ?? '';
  return MIME_TO_EXT[baseMime] ?? '';
}

/**
 * Get MIME type from file extension
 */
export function extensionToMime(extension: string): string {
  const ext = extension.startsWith('.') ? extension.toLowerCase() : `.${extension.toLowerCase()}`;
  return EXT_TO_MIME[ext] ?? 'application/octet-stream';
}

/**
 * Get ResourceType from file extension
 */
export function extensionToType(extension: string): ResourceType {
  const ext = extension.startsWith('.') ? extension.toLowerCase() : `.${extension.toLowerCase()}`;
  return EXT_TO_TYPE[ext] ?? 'other';
}

/**
 * Get ResourceType from MIME type
 */
export function mimeToType(mimeType: string): ResourceType {
  const baseMime = mimeType.split(';')[0]?.trim().toLowerCase() ?? '';

  if (baseMime.includes('html')) return 'html';
  if (baseMime.includes('css')) return 'css';
  if (baseMime.includes('javascript') || baseMime.includes('ecmascript')) return 'javascript';
  if (baseMime.startsWith('image/')) return 'image';
  if (baseMime.startsWith('font/') || baseMime.includes('font')) return 'font';
  if (baseMime.startsWith('video/')) return 'video';
  if (baseMime.startsWith('audio/')) return 'audio';
  if (baseMime.includes('pdf')) return 'document';
  if (baseMime.includes('manifest') || baseMime.includes('json')) return 'manifest';

  return 'other';
}

/**
 * Check if content type is textual (should be processed for links)
 */
export function isTextContent(mimeType: string): boolean {
  const baseMime = mimeType.split(';')[0]?.trim().toLowerCase() ?? '';
  return (
    baseMime.startsWith('text/') ||
    baseMime.includes('javascript') ||
    baseMime.includes('json') ||
    baseMime.includes('xml') ||
    baseMime.includes('+xml') ||
    baseMime.includes('+json')
  );
}

/**
 * Check if extension is supported for processing
 */
export function isSupportedExtension(extension: string): boolean {
  const ext = extension.startsWith('.') ? extension.toLowerCase() : `.${extension.toLowerCase()}`;
  return ext in EXT_TO_TYPE;
}

File: src/util/mime.cpp
#include "surl/util/mime.hpp"

#include "surl/util/strings.hpp"

#include <unordered_map>

namespace surl {
namespace {

struct MimeEntry {
    const char* extension;
    const char* content_type;
    ResourceKind kind;
};

// Ordered so that the first match for a content type is the canonical extension.
constexpr MimeEntry kMimeTable[] = {
    {".html", "text/html", ResourceKind::Html},
    {".htm", "text/html", ResourceKind::Html},
    {".xhtml", "application/xhtml+xml", ResourceKind::Html},
    {".css", "text/css", ResourceKind::Css},
    {".js", "text/javascript", ResourceKind::JavaScript},
    {".mjs", "text/javascript", ResourceKind::JavaScript},
    {".cjs", "text/javascript", ResourceKind::JavaScript},
    {".json", "application/json", ResourceKind::Other},
    {".map", "application/json", ResourceKind::Other},
    {".webmanifest", "application/manifest+json", ResourceKind::Manifest},
    {".xml", "application/xml", ResourceKind::Other},
    {".txt", "text/plain", ResourceKind::Other},
    {".svg", "image/svg+xml", ResourceKind::Image},
    {".png", "image/png", ResourceKind::Image},
    {".jpg", "image/jpeg", ResourceKind::Image},
    {".jpeg", "image/jpeg", ResourceKind::Image},
    {".gif", "image/gif", ResourceKind::Image},
    {".webp", "image/webp", ResourceKind::Image},
    {".avif", "image/avif", ResourceKind::Image},
    {".bmp", "image/bmp", ResourceKind::Image},
    {".ico", "image/x-icon", ResourceKind::Image},
    {".woff", "font/woff", ResourceKind::Font},
    {".woff2", "font/woff2", ResourceKind::Font},
    {".ttf", "font/ttf", ResourceKind::Font},
    {".otf", "font/otf", ResourceKind::Font},
    {".eot", "application/vnd.ms-fontobject", ResourceKind::Font},
    {".mp4", "video/mp4", ResourceKind::Media},
    {".webm", "video/webm", ResourceKind::Media},
    {".ogv", "video/ogg", ResourceKind::Media},
    {".mp3", "audio/mpeg", ResourceKind::Media},
    {".ogg", "audio/ogg", ResourceKind::Media},
    {".wav", "audio/wav", ResourceKind::Media},
    {".m4a", "audio/mp4", ResourceKind::Media},
    {".flac", "audio/flac", ResourceKind::Media},
    {".pdf", "application/pdf", ResourceKind::Other},
    {".zip", "application/zip", ResourceKind::Other},
    {".wasm", "application/wasm", ResourceKind::Other},
};

std::string normalize_extension(std::string_view extension) {
    std::string ext = to_lower(trim(extension));
    if (!ext.empty() && ext.front() != '.') ext.insert(ext.begin(), '.');
    return ext;
}

} // namespace

std::string content_type_essence(std::string_view content_type) {
    const std::size_t semicolon = content_type.find(';');
    std::string_view head =
        (semicolon == std::string_view::npos) ? content_type : content_type.substr(0, semicolon);
    return to_lower(trim(head));
}

ResourceKind kind_from_content_type(std::string_view content_type) {
    const std::string essence = content_type_essence(content_type);
    if (essence.empty()) return ResourceKind::Other;

    if (essence == "text/html" || essence == "application/xhtml+xml") return ResourceKind::Html;
    if (essence == "text/css") return ResourceKind::Css;
    if (essence == "text/javascript" || essence == "application/javascript" ||
        essence == "application/x-javascript" || essence == "text/ecmascript" ||
        essence == "application/ecmascript") {
        return ResourceKind::JavaScript;
    }
    if (essence == "application/manifest+json") return ResourceKind::Manifest;

    if (istarts_with(essence, "image/")) return ResourceKind::Image;
    if (istarts_with(essence, "font/")) return ResourceKind::Font;
    if (istarts_with(essence, "video/") || istarts_with(essence, "audio/")) {
        return ResourceKind::Media;
    }
    if (essence == "application/font-woff" || essence == "application/x-font-ttf" ||
        essence == "application/vnd.ms-fontobject") {
        return ResourceKind::Font;
    }
    return ResourceKind::Other;
}

ResourceKind kind_from_extension(std::string_view extension) {
    const std::string ext = normalize_extension(extension);
    for (const MimeEntry& entry : kMimeTable) {
        if (ext == entry.extension) return entry.kind;
    }
    return ResourceKind::Other;
}

std::string extension_for_content_type(std::string_view content_type) {
    const std::string essence = content_type_essence(content_type);
    if (essence.empty()) return "";
    for (const MimeEntry& entry : kMimeTable) {
        if (essence == entry.content_type) return entry.extension;
    }
    // A few aliases that are common enough to be worth handling.
    if (essence == "application/javascript" || essence == "application/x-javascript") {
        return ".js";
    }
    if (essence == "image/jpg") return ".jpg";
    if (essence == "image/vnd.microsoft.icon") return ".ico";
    return "";
}

std::string content_type_for_extension(std::string_view extension) {
    const std::string ext = normalize_extension(extension);
    for (const MimeEntry& entry : kMimeTable) {
        if (ext == entry.extension) {
            const ResourceKind kind = entry.kind;
            std::string type = entry.content_type;
            // Text formats need an explicit charset or browsers guess wrong.
            if (kind == ResourceKind::Html || kind == ResourceKind::Css ||
                kind == ResourceKind::JavaScript || ext == ".json" || ext == ".txt" ||
                ext == ".xml" || ext == ".map" || ext == ".webmanifest") {
                type += "; charset=utf-8";
            }
            return type;
        }
    }
    return "application/octet-stream";
}

const char* kind_name(ResourceKind kind) {
    switch (kind) {
    case ResourceKind::Html: return "html";
    case ResourceKind::Css: return "css";
    case ResourceKind::JavaScript: return "js";
    case ResourceKind::Image: return "image";
    case ResourceKind::Font: return "font";
    case ResourceKind::Media: return "media";
    case ResourceKind::Manifest: return "manifest";
    case ResourceKind::Other: break;
    }
    return "other";
}

} // namespace surl
